mos-platform/common/lib/new.cc

```diff
@@ -0,0 +1,410 @@
+    m_used = sizeof(block_node);
+    m_free = m_head->m_block.m_size;
+  }
+
+  struct iterator {
+  public:
+    iterator() = default;
+    iterator(block_node *ptr) noexcept : m_ptr{ptr} {}
+
+    block &operator*() const noexcept { return m_ptr->m_block; }
+
+    iterator &operator++() noexcept {
+      m_ptr = static_cast<block_node *>(m_ptr->m_next);
+      return *this;
+    }
+
+    bool operator==(const iterator &rhs) const noexcept {
+      return m_ptr == rhs.m_ptr;
+    }
+
+    bool operator!=(const iterator &rhs) const noexcept {
+      return !this->operator==(rhs);
+    }
+
+  private:
+    block_node *m_ptr = nullptr;
+  };
+
+  iterator begin() { return m_head; }
+
+  iterator end() { return iterator{}; }
+
+  static_assert(sizeof(list_node) ==
+                    sizeof(list_node::m_next) + sizeof(list_node::m_prev),
+                "unexpected padding");
+  static_assert(sizeof(block_node) == sizeof(list_node) + sizeof(std::size_t),
+                "unexpected padding");
+
+  std::byte *split_block(block *freeblock, std::size_t size) noexcept {
+    if (!freeblock) {
+      return nullptr;
+    }
+
+    if ((freeblock->m_size - size) >= blocklist::MIN_ALLOC_SIZE) {
+      const auto new_blk =
+          new (freeblock->data() + size) block_node{freeblock->m_size - size};
+      freeblock->m_size = size;
+
+      const auto freeblock_node_ptr = block_node::get_node(*freeblock);
+      {
+        const auto next_block_ptr = new_blk->m_next =
+            freeblock_node_ptr->m_next;
+        if (next_block_ptr) {
+          next_block_ptr->m_prev = new_blk;
+        }
+      }
+
+      const auto prev_block_ptr = new_blk->m_prev = freeblock_node_ptr->m_prev;
+      if (prev_block_ptr) {
+        prev_block_ptr->m_next = new_blk;
+      } else {
+        m_head = new_blk;
+      }
+
+      // A new node in free list was allocated. Remove it from the total bytes
+      // free.
+      m_used += sizeof(block_node);
+      m_free -= sizeof(block_node);
+    } else {
+      // Drop the block from the free list entirely.
+      const auto freeblock_node_ptr = block_node::get_node(*freeblock);
+      const auto prev_block_ptr = freeblock_node_ptr->m_prev;
+      const auto next_block_ptr = freeblock_node_ptr->m_next;
+
+      if (prev_block_ptr) {
+        prev_block_ptr->m_next = next_block_ptr;
+      } else {
+        m_head = static_cast<block_node *>(next_block_ptr);
+      }
+
+      if (next_block_ptr) {
+        next_block_ptr->m_prev = prev_block_ptr;
+      }
+    }
+
+    // Update stats for actual size allocated.
+    m_used += freeblock->m_size;
+    m_free -= freeblock->m_size;
+    return freeblock->data();
+  }
+
+  block *find_first_fit(std::size_t sz) noexcept {
+    for (auto &block : *this) {
+      if (block.m_size >= (sz + sizeof(block_node))) {
+        return &block;
+      }
+    }
+
+    return nullptr;
+  }
+
+  void free_block(block *const block) noexcept {
```

Comment:
Add a TODO  comment for each of these operations that are O(n). This PR is for a simple heap implementation, but simple is a very relative term, and it's time complexity is an order of magnitude worse than typical, even for far simpler approaches than a full dlmalloc.

Reply:
Commentary added. If someone comes up with an implementation for MOS that is really an order of magnitude faster in terms of real run time, I welcome them to show it.  

Keep in mind that the (n) in these operations is low: in practical terms it's in the 100's at most, and absolute maximum (n) is only 4096.